core/resource/store/store.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package store
+
+import (
+	"context"
+	"io"
+
+	"github.com/juju/utils/v4/hash"
+)
+
+// ResourceStore provides a list of methods necessary for interacting with
+// a store for the resource.
+type ResourceStore interface {
+	// Get returns an io.ReadCloser for a resource in the resource store.
+	Get(
+		ctx context.Context,
+		storageKey string,
+	) (r io.ReadCloser, size int64, err error)
+
+	// Put stores data from io.Reader in the resource store using the storage
+	// key.
+	Put(
+		ctx context.Context,
+		storageKey string,
+		r io.Reader,
+		size int64,
+		fingerprint Fingerprint,
+	) (UUID, error)
+
+	// Remove removes a resource from storage.
+	Remove(
+		ctx context.Context,
+		storageKey string,
+	) error
+}
+
+// Fingerprint represents the unique fingerprint value of a resource's data.
+type Fingerprint struct {
+	hash.Fingerprint
+}
+
+// NewFingerprint returns a resource store Fingerprint for the given
+// hash Fingerprint.
+func NewFingerprint(f hash.Fingerprint) Fingerprint {
+	return Fingerprint{f}
+}
+
+// UUID is the UUID of the stored blob in the database, this can
+// be used for adding referential integrity from the resource to the stored
+// blob. This can be an object store metadata UUID or a container image metadata
+// storage key.
+type UUID string
```

Comment:
This should be dropped. Just use the objectstore.UUID here. By having another type, it suggests that this is something it's not. This will be and will only ever be an objectstore.UUID.

Reply:
Thats not true. This is either the primary key from the container image resource store table, or the object store UUID, it can be either, that is why there is a type for it. This is explained in the comment above.

Reply:
Then this shouldn't be a type string. You're essentially dropping information about the origin of the type. If I were reading this code in the future, the fact that you're masking it means I have to understand that. That's hidden control flow.

Reply:
@Aflynn50 please resolve in a new PR.